contracts/BonusRewards.sol

```diff
@@ -3,11 +3,363 @@
+        if (bonus.startTime > block.timestamp) {
+          // only honor new start time, if program has not started
+          if (_startTime > block.timestamp) {
+            bonus.startTime = _startTime;
+          }
+          bonus.endTime = bonus.remBonus * WEEK / _weeklyRewards  + bonus.startTime;
+        } else {
+          uint256 remBonusToDistribute = (bonus.endTime -  block.timestamp) * bonus.weeklyRewards / WEEK;
+          bonus.endTime = remBonusToDistribute * WEEK / _weeklyRewards + block.timestamp;
+        }
+        bonus.weeklyRewards = _weeklyRewards;
+      }
+    }
+  }
+
+  /// @notice extend the current bonus program, the program has to be active (endTime is in the future)
+  function extendBonus(
+    address _lpToken,
+    uint256 _poolBonusId,
+    address _bonusTokenAddr,
+    uint256 _transferAmount
+  ) external override nonReentrant notPaused {
+    require(_isAuthorized(msg.sender, allowedTokenAuthorizers[_lpToken][_bonusTokenAddr]), "BonusRewards: not authorized caller");
+
+    Bonus memory bonus = pools[_lpToken].bonuses[_poolBonusId];
+    require(bonus.bonusTokenAddr == _bonusTokenAddr, "BonusRewards: bonus and id dont match");
+    require(bonus.endTime > block.timestamp, "BonusRewards: bonus program ended, please start a new one");
+
+    IERC20 bonusTokenAddr = IERC20(_bonusTokenAddr);
+    uint256 balanceBefore = bonusTokenAddr.balanceOf(address(this));
+    bonusTokenAddr.safeTransferFrom(msg.sender, address(this), _transferAmount);
+    uint256 received = bonusTokenAddr.balanceOf(address(this)) - balanceBefore;
+    // endTime is based on how much tokens transfered v.s. planned weekly rewards
+    uint256 endTime = received * WEEK / bonus.weeklyRewards + bonus.endTime;
+
+    pools[_lpToken].bonuses[_poolBonusId].endTime = endTime;
+    pools[_lpToken].bonuses[_poolBonusId].remBonus = bonus.remBonus + received;
+  }
+
+  /// @notice add pools and authorizers to add bonus tokens for pools, combine two calls into one. Only reason we add pools is when bonus tokens will be added
+  function addPoolsAndAllowBonus(
+    address[] calldata _lpTokens,
+    address[] calldata _bonusTokenAddrs,
+    address[] calldata _authorizers
+  ) external override onlyOwner notPaused {
+    // add pools
+    for (uint256 i = 0; i < _lpTokens.length; i++) {
+      address _lpToken = _lpTokens[i];
+      Pool memory pool = pools[_lpToken];
+      if (pool.lastUpdatedAt == 0) {
+        pools[_lpToken].lastUpdatedAt = block.timestamp;
+        poolList.push(_lpToken);
+      }
+
+      // add bonus tokens and their authorizers (who are allowed to add the token to pool)
+      for (uint256 j = 0; j < _bonusTokenAddrs.length; j++) {
+        address _bonusTokenAddr = _bonusTokenAddrs[j];
+        require(pools[_bonusTokenAddr].lastUpdatedAt == 0, "BonusRewards: lpToken, not allowed");
+        allowedTokenAuthorizers[_lpToken][_bonusTokenAddr] = _authorizers;
+        if (bonusTokenAddrMap[_bonusTokenAddr] == 0) {
```

Comment:
Is this check needed?

Reply:
It is used in collectDust but I am not sure it is needed as we check: bonus.bonusTokenAddr == _token. @crypto-pumpkin ?

Reply:
It saves a write to storage action. It is possible to have the same bonus token on different pools.